File: StockChess/StockChessCS/Services/StockfishService.cs
using System;
using System.IO;
using System.Reactive.Linq;
using System.Diagnostics;
using StockChessCS.Helpers;
using StockChessCS.Interfaces;

namespace StockChessCS.Services
{
    public class StockfishService : IEngineService
    {

        private StreamReader strmReader;
        private StreamWriter strmWriter;
        private Process engineProcess;

        private IDisposable engineListener;
        public event Action<string> EngineMessage;	

        public void SendCommand(string command)
        {
            if (strmWriter != null && command != UciCommands.uci)
            {
                strmWriter.WriteLine(command);
            }
        }

        public void StopEngine()
        {
            if (engineProcess != null & !engineProcess.HasExited)
            {
                engineListener.Dispose();
                strmReader.Close();
                strmWriter.Close();                
            }
        }

        public void StartEngine()
        {
            FileInfo engine = new FileInfo(Path.Combine(Environment.CurrentDirectory, "stockfish_8_x64.exe"));
            if (engine.Exists && engine.Extension == ".exe")
            {
                engineProcess = new Process();
                engineProcess.StartInfo.FileName = engine.FullName;
                engineProcess.StartInfo.UseShellExecute = false;
                engineProcess.StartInfo.RedirectStandardInput = true;
                engineProcess.StartInfo.RedirectStandardOutput = true;
                engineProcess.StartInfo.RedirectStandardError = true;
                engineProcess.StartInfo.CreateNoWindow = true;

                engineProcess.Start();

                strmWriter = engineProcess.StandardInput;
                strmReader = engineProcess.StandardOutput;

                engineListener = Observable.Timer(TimeSpan.Zero, TimeSpan.FromMilliseconds(1)).Subscribe(s => ReadEngineMessages());

                strmWriter.WriteLine(UciCommands.uci);
                strmWriter.WriteLine(UciCommands.isready);
                strmWriter.WriteLine(UciCommands.ucinewgame);
            }
            else
            {
                throw new FileNotFoundException();
            }
        }

        private void ReadEngineMessages()
        {
            var message = strmReader.ReadLine();
            if (message != string.Empty)
            {
                EngineMessage?.Invoke(message);
            }
        }
    }
}

File: StockChess/StockChess/Services/StockfishService.vb
Imports System.IO
Imports System.Reactive.Linq

Public Class StockfishService
    Implements IEngineService

    Private strmReader As StreamReader
    Private strmWriter As StreamWriter
    Private engineProcess As Process
    Private engineListener As IDisposable

    Public Event EngineMessage(message As String) Implements IEngineService.EngineMessage

    Public Sub SendCommand(command As String) Implements IEngineService.SendCommand
        If strmWriter IsNot Nothing AndAlso command <> UciCommands.uci Then
            strmWriter.WriteLine(command)
        End If
    End Sub

    Public Sub StopEngine() Implements IEngineService.StopEngine
        If engineProcess IsNot Nothing And Not engineProcess.HasExited Then
            engineListener.Dispose()
            strmReader.Close()
            strmWriter.Close()
        End If
    End Sub

    Public Sub StartEngine() Implements IEngineService.StartEngine
        Dim engine As New FileInfo(Path.Combine(Environment.CurrentDirectory, "stockfish_8_x64.exe"))
        If engine.Exists AndAlso engine.Extension = ".exe" Then
            engineProcess = New Process
            engineProcess.StartInfo.FileName = engine.FullName
            engineProcess.StartInfo.UseShellExecute = False
            engineProcess.StartInfo.RedirectStandardInput = True
            engineProcess.StartInfo.RedirectStandardOutput = True
            engineProcess.StartInfo.RedirectStandardError = True
            engineProcess.StartInfo.CreateNoWindow = True

            engineProcess.Start()

            strmWriter = engineProcess.StandardInput
            strmReader = engineProcess.StandardOutput

            engineListener = Observable.Timer(TimeSpan.Zero, TimeSpan.FromMilliseconds(1)).Subscribe(Sub() ReadEngineMessages())

            strmWriter.WriteLine(UciCommands.uci)
            strmWriter.WriteLine(UciCommands.isready)
            strmWriter.WriteLine(UciCommands.ucinewgame)
        Else
            Throw New FileNotFoundException
        End If
    End Sub

    Private Sub ReadEngineMessages()
        Dim message = strmReader.ReadLine()
        If message <> String.Empty Then
            RaiseEvent EngineMessage(message)
        End If
    End Sub
End Class